src/fsharp/IlxGen.fs

```diff
@@ -4107,18 +4107,45 @@ and GenIntegerForLoop cenv cgbuf eenv (spFor, spTo, v, e1, dir, e2, loopBody, m)
 
 and GenWhileLoop cenv cgbuf eenv (spWhile, condExpr, bodyExpr, m) sequel =
     let eenv = SetIsInLoop true eenv
-    let finish = CG.GenerateDelayMark cgbuf "while_finish"
 
-    match spWhile with
-    | DebugPointAtWhile.Yes spStart -> CG.EmitDebugPoint cgbuf spStart
-    | DebugPointAtWhile.No -> ()
+    // jmp test; body; test; if testPassed then jmp body else finish
+    // 
+    // This is a pattern recognized by the JIT and it results in the most efficient assembly.
+    if cgbuf.GetCurrentStack().IsEmpty then
```

Comment:
Will it be a lot of work to have two separate tests in `FSharp.Compiler.ComponentTests\EmittedIL` which will verify both branches, to see if we have expected codegen?
Thanks.

Reply:
2 such tests are in the new WhileLoops.fs. It's in ComponentTests, not as a baseline file though.

Reply:
> 2 such tests are in the new WhileLoops.fs

Oh, apologies, it seems that GH just collapsed the file and I missed it.
Thanks!